### PR TITLE
fix(ZNTA-1740): ensure UTF-8 when parsing TMX segments

### PR DESCRIPTION
**Pull request by seanf**
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-1740

TMX import was previously broken for non-Unicode locales. This change ensures
TMX will be imported as Unicode, and fixes existing translation memories. TMX
importing should also be a bit faster.

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines#new-technologies-and-dependencies)?

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/208)

### CODE REVIEW

File: docs/user-guide/system-admin/configuration/installation.md [hunk 1]
```diff
@@ -95,6 +95,31 @@ This setting will cause Zanata to output CORS headers for REST requests, includi
 `Access-Control-Allow-Origin` (if Origin is in the whitelist above),
 `Access-Control-Allow-Credentials: true` and
 `Access-Control-Allow-Methods: PUT,POST,DELETE,GET,OPTIONS`.
+
+
+## Redirecting HTTP to HTTPS
```
Comment (carlosmunoz) [inline docs/user-guide/system-admin/configuration/installation.md]: Why is this part of this pull request?

File: server/zanata-war/src/main/java/org/zanata/liquibase/custom/RebuildTMXPlainText.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import liquibase.exception.SetupException;
+import liquibase.exception.ValidationErrors;
+import liquibase.logging.LogFactory;
+import liquibase.logging.Logger;
+import liquibase.resource.ResourceAccessor;
+
+/**
+ * @author Sean Flanigan <a href="mailto:sflaniga@redhat.com">sflaniga@redhat.com</a>
+ */
+public class RebuildTMXPlainText implements CustomTaskChange {
```
Comment (carlosmunoz) [inline server/zanata-war/src/main/java/org/zanata/liquibase/custom/RebuildTMXPlainText.java]: Lets add a class description for the custom change.
I think I know what it does, but others might not.

File: server/zanata-war/src/main/java/org/zanata/servlet/UrlRewriteConfig.java [hunk 3]
```diff
@@ -25,19 +28,41 @@
...
+            Boolean.getBoolean("zanata.redirectHttpToHttps");
+
     @Override
     public Configuration getConfiguration(final ServletContext context) {
         String contextPath = context.getContextPath();
         // NB: inbound rules are processed in order, outbound rules in reverse order (as of Rewrite 3.0.0.Alpha1)
-        return ConfigurationBuilder.begin()
-
+        ConfigurationBuilder builder = ConfigurationBuilder.begin();
+        if (REDIRECT_TO_HTTPS) {
```
Comment (carlosmunoz) [inline server/zanata-war/src/main/java/org/zanata/servlet/UrlRewriteConfig.java]: Hmmm.... again, this looks like it belongs in a separate PR

File: server/zanata-war/src/test/java/org/zanata/tmx/TMXParserTest.java [hunk 4]
```diff
@@ -208,6 +212,14 @@ public void parseTMX() throws Exception {
...
             assertThat(tu.getTransUnitVariants().size(), greaterThan(0));
         }
+
+        Optional<String> seg = tm.getTranslationUnits().stream()
+                .filter(tu -> tu.getTransUnitId() != null && tu.getTransUnitId().equals("tuid2"))
+                .map(TransMemoryUnit::getTransUnitVariants)
+                .map(map ->  map.get("ja"))
+                .map(TransMemoryUnitVariant::getPlainTextSegment)
+                .findAny();
+        assertThat(seg.orElseGet(null), containsString("を選択します"));
```
Comment (carlosmunoz) [inline server/zanata-war/src/test/java/org/zanata/tmx/TMXParserTest.java]: Not sure we need this assertion here. It was not detecting the issue last I remembered.

File: server/zanata-war/src/test/resources/tmx/default-valid-tm.tmx [hunk 5]
```diff
@@ -39,7 +39,7 @@ o-tmf="zanata.org"
         <seg>Questo è il primo elemento.</seg>
     </tuv>
   </tu>
-  <tu>
+  <tu tuid="tuid2">
```
Comment (carlosmunoz) [inline server/zanata-war/src/test/resources/tmx/default-valid-tm.tmx]: Same here... doesn't hurt to have it, but I don't think we need this change.

File: docs/user-guide/system-admin/configuration/installation.md (hunk 1 above)
Comment (seanf) [inline docs/user-guide/system-admin/configuration/installation.md]: It's not really. I built this on top of the other branch, which hasn't been merged yet.

File: server/zanata-war/src/test/java/org/zanata/tmx/TMXParserTest.java (hunk 4 above)
Comment (seanf) [inline server/zanata-war/src/test/java/org/zanata/tmx/TMXParserTest.java]: It didn't detect it in this case, but I still think it's a useful test.

If we were to run our unit tests in a POSIX/ASCII environment (and perhaps we should), I think it might actually have detected the problem.

File: server/zanata-model/src/main/java/org/zanata/model/tm/TransMemoryUnitVariant.java [hunk 6]
```diff
@@ -86,8 +86,8 @@ public void setTaggedSegment(String taggedSegment) {
     }
 
     private void updatePlainTextSegment() {
-        this.plainTextSegment = OkapiUtil.removeFormattingMarkup(taggedSegment);
         updatePlainTextSegmentHash();
+        this.plainTextSegment = TMXUtil.removeFormattingMarkup(taggedSegment);
```
Comment (seanf) [inline server/zanata-model/src/main/java/org/zanata/model/tm/TransMemoryUnitVariant.java]: A bit of accidental mutation testing. Fortunately the unit tests caught it.

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by seanf** [comment]
Changing the base branch to `master` and back to `release` _did_ stop github from listing the two commits which were merged as part of https://github.com/zanata/zanata-platform/pull/196, but it triggered two builds, whereas a force push would only have produced one. I was hoping to avoid _any_ useless rebuilds. So much for that idea.
